# Практическая работа №6 «JavaScript. Математические функции. Строковые функции. Функции для массивов»

# «Математические функции»

# Задания

## Задание 1: «Работа с %»

Задача 1.1 Даны переменные a = 10 и b = 3. Найти остаток от деления a на b.

In [9]:
a = 10
b = 3
console.log(a % b)

1


Задача 1.2 Даны переменные a и b. Проверить, что a делится без остатка на b. Если это так - выведите 'Делится' и результат деления, иначе вывести 'Делится с остатком' и остаток от деления.

In [1]:
check = (a, b) => {
    a % b == 0 ? console.log(`Делится = ${a / b}`) 
    : console.log(`Делится с остатком = ${a % b}`)
}

check(5, 2)
check(10, 5)

Делится с остатком = 1
Делится = 2


## Задание 2: «Работа со степенью и корнем»

### Для решения задач данного блока вам понадобятся следующие методы: Math.pow, Math.sqrt.

Задача 2.1 Возвести 2 в 10 степень. Результат записать в переменную st.

In [11]:
st = Math.pow(2, 10)
console.log(st)

1024


Задача 2.2 Найти квадратный корень из 245.

In [12]:
console.log(Math.sqrt(245))

15.652475842498529


Задача 2.3 Дан массив с элементами 4, 2, 5, 19, 13, 0, 10. Найти квадратный корень из суммы кубов его элементов. Для решения воспользоваться циклом for

In [13]:
arr = [4, 2, 5, 19, 13, 0, 10]
sum = 0

for(i in arr)
    sum += Math.pow(arr[i], 3)

console.log(Math.sqrt(sum))

101.25709851659784


## Задание 3: «Работа с функциями округления»

### Для решения задач данного блока вам понадобятся следующие функции: Math.round, Math.ceil, Math.floor, toFixed, toPrecision.

Задача 3.1 Найти квадратный корень из 379. Результат округлить до целых, до десятых, до сотых.

In [23]:
console.log(Math.round(Math.sqrt(379)))
console.log(Math.sqrt(379).toPrecision(3))
console.log(Math.sqrt(379).toPrecision(4))

19
19.5
19.47


Задача 3.2 Найти квадратный корень из 587. Округлить результат в большую и меньшую стороны, записать результаты округления в объект с ключами 'floor' и 'ceil'.

In [25]:
console.log(`Floor = ${Math.floor(Math.sqrt(587))}`)
console.log(`Ceil = ${Math.ceil(Math.sqrt(587))}`)

Floor = 24
Ceil = 25


## Задание 4: «Нахождение максимального и минимального числа»

### Для решения задачи данного блока вам понадобятся следующие методы: Math.max, Math.min.

Задача 4.1 Даны числа 4, -2, 5, 19, -130, 0, 10. Найти минимальное и максимальное число.

In [4]:
console.log(`Min = ${Math.min(4, -2, 5, 19, -130, 0, 10)}
Max = ${Math.max(4, -2, 5, 19, -130, 0, 10)}`)

Min = -130
Max = 19


## Задание 5: «Работа с рандомом»

### Для решения задач данного блока вам понадобятся следующие методы: Math.random.

Задача 5.1 Вывести на экран случайное целое число от 1 до 100.

In [30]:
console.log(Math.floor(Math.random() * 100) + 1)

17


Задача 5.2 Заполнить массив 10-ю случайными целыми числами. (Подсказка: нужно воспользоваться циклами for или while). 

In [33]:
for(var i = 0; i < 10; i++)
    arr[i] = Math.floor(Math.random() * 100) + 1

console.log(arr)

[
  95, 96, 77, 25, 100,
  70, 68, 25, 32,  86
]


## Задание 6: «Работа с модулем»

### Для решения задач данного блока вам понадобятся следующие методы: Math.abs.

Задача 6.1 Даны переменные a и b. Найти модуль разности a и b. Проверить работу скрипта для различных a и b.

In [36]:
res = (a, b) => console.log(Math.abs(a - b))

res(10, 3)
res(5, 6)

7
1


Задача 6.2 Даны переменные a и b. Отнять от a переменную b и результат присвоить переменной c. Сделать так, чтобы в любом случае в переменную c записалось положительное значение. Проверить работу скрипта при a и b, равных соответственно 3 и 5, 6 и 1.

In [37]:
res = (a, b) => Math.abs(a - b)

c = res(3, 5)
console.log(c)

c = res(6, 1)
console.log(c)

2
5


## Задание 7

Задача 7.1 Дан массив arr. Найти среднее арифметическое его элементов. Проверить задачу на массиве с элементами 12, 15, 20, 25, 59, 79.

In [38]:
arr = [12, 15, 20, 25, 59, 79]
console.log((arr.reduce((a, b) => a + b, 0)) / [1,2,3,4,5].length)

42


Задача 7.2 Написать скрипт, который будет находить факториал числа. Факториал (обозначается !) - это произведение (умножение) всех целых чисел, меньше данного, и его самого. Например, 4! = 1*2*3*4.

In [40]:
fc = n => n === 0 || n === 1 ? 1 : n * fc(n - 1)

console.log(fc(4))

24


# «Строковые функции»
# Задания

## Задание 8: «Работа с регистром символов»

### Для решения задач данного блока вам понадобятся следующие методы: toUpperCase, toLowerCase.

Задача 8.1 Дана строка 'js'. Сделать из нее строку 'JS'.

In [42]:
str = 'js'
console.log(str.toUpperCase())

JS


Задача 8.2 Дана строка 'JS'. Сделать из нее строку 'js'.

In [43]:
str = 'js'
console.log(str.toLowerCase())

js


## Задание 9: «Работа с length, substr, substring, slice. Работа с indexOf»

Задача 9.1 Дана строка 'я учу javascript!'. Найти количество символов в этой строке.

In [44]:
str = 'я учу javascript!'
console.log(str.length)

17


Задача 9.2 Дана строка 'я учу javascript!'. Вырезать из нее слово 'учу' и слово 'javascript' тремя разными способами (через substr, substring, slice).

In [66]:
str1 = 'я учу javascript!'
str2 = 'я учу javascript!'
str3 = 'я учу javascript!'

str1 = str1.substr(2, 4)
str2 = str2.substring(6, 16)
str3 = str3.slice(2, 5)

console.log(str1)
console.log(str2)
console.log(str3)

учу 
javascript
учу


Задача 9.3 Дана строка 'я учу javascript!'. Найти позицию подстроки 'учу'.

In [67]:
str = 'я учу javascript!'
console.log(str.indexOf('учу'))

2


Задача 9.4 Дана переменная str, в которой хранится какой-либо текст. Реализовать обрезание длинного текста по следующему принципу: если количество символов этого текста больше заданного в переменной n, то в переменную result записать первые n символов строки str и добавить в конец троеточие '...'. В противном случае в переменную result записать содержимое переменной str.

In [74]:
sl = (text, n) => n < text.length ? result = text.substr(0, n) + "..." : result = text

sl("The quick brown fox jumps over the lazy dog", 11)
console.log(result)

sl("The quick brown fox jumps over the lazy dog", 50)
console.log(result)

The quick b...
The quick brown fox jumps over the lazy dog


## Задание 10: «Работа с replace»

Задача 10.1 Дана строка 'Я-учу-javascript!'. Заменить все дефисы на '!' с помощью глобального поиска и замены.

In [78]:
console.log('Я-учу-javascript!'.replace(/-/g, '!'))

Я!учу!javascript!


## Задание 11: «Работа с split»

Задача 11.1 Дана строка 'я учу javascript!'. С помощью метода split записать каждое слово этой строки в отдельный элемент массива.

In [81]:
arr = 'я учу javascript!'.split(' ')
console.log(arr)

[ 'я', 'учу', 'javascript!' ]


Задача 11.2 Дана строка 'я учу javascript!'. С помощью метода split записать каждый символ этой строки в отдельный элемент массива.

In [82]:
arr = 'я учу javascript!'.split('')
console.log(arr)

[
  'я', ' ', 'у', 'ч', 'у',
  ' ', 'j', 'a', 'v', 'a',
  's', 'c', 'r', 'i', 'p',
  't', '!'
]


Задача 11.3 В переменной date лежит дата в формате '2020-11-18'. Преобразовать эту дату в формат '18.11.2020'.

In [85]:
date = '2020-11-18'
console.log(`${date.slice(8, 10)}.${date.slice(5, 7)}.${date.slice(0, 4)}`)

18.11.2020


## Задание 12: «Работа с join»

Задача 12.1 Дан массив ['я', 'учу', 'javascript', '!']. С помощью метода join преобразовать массив в строку 'я+учу+javascript+!'.

In [86]:
arr = ['я', 'учу', 'javascript', '!']
console.log(arr.join('+'))

я+учу+javascript+!


# «Функции для массивов»
# Задания

## Задание 13: «Работа с concat»

Даны два массива: [1, 2, 3] и [4, 5, 6]. Объединить их вместе.

In [87]:
a = [1, 2, 3]
b = [4, 5, 6]
console.log(a.concat(b))

[ 1, 2, 3, 4, 5, 6 ]


## Задание 14: «Работа с reverse»

Дан массив [1, 2, 3]. Сделать из него массив [3, 2, 1].

In [88]:
arr = [1, 2, 3]
console.log(arr.reverse())

[ 3, 2, 1 ]


## Задание 15: «Работа с push, unshift»

Задача 15.1 Дан массив [1, 2, 3]. Добавить ему в конец элементы 4, 5, 6.

In [89]:
arr = [1, 2, 3]
arr.push(4, 5, 6)
console.log(arr)

[ 1, 2, 3, 4, 5, 6 ]


Задача 15.2 Дан массив [1, 2, 3]. Добавить ему в начало элементы 4, 5, 6.

In [91]:
arr = [1, 2, 3]
arr.unshift(4, 5, 6)
console.log(arr)

[ 4, 5, 6, 1, 2, 3 ]


## Задание 16: «Работа с shift, pop»

Задача 16.1 Дан массив ['js', 'css', 'jq']. Вывести на экран первый элемент.

In [93]:
arr = ['js', 'css', 'jq']
console.log(arr.shift())

js


Задача 16.2 Дан массив ['js', 'css', 'jq']. Вывести на экран последний элемент.

In [94]:
arr = ['js', 'css', 'jq']
console.log(arr.pop())

jq


## Задание 17: «Работа со slice»

Задача 17.1 Дан массив [1, 2, 3, 4, 5]. С помощью метода slice записать в новый элементы [1, 2, 3].

In [96]:
arr = [1, 2, 3, 4, 5]
console.log(arr.slice(0, 3))

[ 1, 2, 3 ]


Задача 17.2 Дан массив [1, 2, 3, 4, 5]. С помощью метода slice записать в новый элементы [4, 5].

In [102]:
arr = [1, 2, 3, 4, 5]
console.log(arr.slice(3, 5))

[ 4, 5 ]


## Задание 18: «Работа со splice»

Задача 18.1 Дан массив [1, 2, 3, 4, 5]. С помощью метода splice преобразовать массив в [1, 4, 5].

In [104]:
arr = [1, 2, 3, 4, 5]
arr.splice(1, 2)
console.log(arr)

[ 1, 4, 5 ]


Задача 18.2 Дан массив [1, 2, 3, 4, 5]. С помощью метода splice записать в новый массив элементы [2, 3, 4].

In [105]:
arr = [1, 2, 3, 4, 5]
narr = arr.splice(1, 3)
console.log(narr)

[ 2, 3, 4 ]


Задача 18.3 Дан массив [1, 2, 3, 4, 5]. С помощью метода splice сделать из него массив [1, 2, 3, 'a', 'b', 'c', 4, 5].

In [107]:
arr = [1, 2, 3, 4, 5]
arr.splice(3, 0, 'a', 'b', 'c')
console.log(arr)

[
  1,   2,   3, 'a',
  'b', 'c', 4, 5
]


Задача 18.4 Дан массив [1, 2, 3, 4, 5]. С помощью метода splice сделать из него массив [1, 'a', 'b', 2, 3, 4, 'c', 5, 'e'].

In [115]:
arr = [1, 2, 3, 4, 5]
arr.splice(1, 0, 'a', 'b')
arr.splice(6, 0, 'c')
arr.splice(8, 0, 'e')
console.log(arr)

[
  1, 'a', 'b', 2,   3,
  4, 'c', 5,   'e'
]


## Задание 19: «Работа с sort»

Дан массив [3, 4, 1, 2, 7]. Отсортировать его.

In [116]:
console.log([3, 4, 1, 2, 7].sort())

[ 1, 2, 3, 4, 7 ]


## Задание 20: «Работа с Object.keys»

Дан объект {js:'test', jq: 'hello', css: 'world'}. Получить массив его ключей.

In [118]:
console.log(Object.keys({js:'test', jq: 'hello', css: 'world'}))

[ 'js', 'jq', 'css' ]
